accounts/abi/event.go

```diff
@@ -57,55 +57,57 @@ func (e Event) tupleUnpack(v interface{}, output []byte) error {
 	}
 
 	var (
-		value = valueOf.Elem()
-		typ   = value.Type()
+		value    = valueOf.Elem()
+		typ      = value.Type()
+		isStruct = true
 	)
-
-	if value.Kind() != reflect.Struct {
-		return fmt.Errorf("abi: cannot unmarshal tuple in to %v", typ)
+	switch value.Kind() {
+	case reflect.Struct:
+	case reflect.Slice, reflect.Array:
+		isStruct = false
+		minLen := countNonIndexedArguments(e.Inputs)
+		if value.Len() < minLen {
+			return fmt.Errorf("abi: insufficient number of elements in the list/array for unpack, want %d, got %d",
+				minLen, value.Len())
+		}
+	default:
+		return fmt.Errorf("abi: cannot unmarshal tuple into %v", typ)
 	}
 
-	j := 0
-	for i := 0; i < len(e.Inputs); i++ {
-		input := e.Inputs[i]
+	i, j := -1, 0
+	for _, input := range e.Inputs {
 		if input.Indexed {
-			// can't read, continue
+			// Indexed arguments are not packed into data
 			continue
 		} else if input.Type.T == ArrayTy {
 			// need to move this up because they read sequentially
 			j += input.Type.Size
 		}
+		i++
 		marshalledValue, err := toGoType((i+j)*32, input.Type, output)
 		if err != nil {
 			return err
 		}
 		reflectValue := reflect.ValueOf(marshalledValue)
 
-		switch value.Kind() {
-		case reflect.Struct:
+		if isStruct {
 			for j := 0; j < typ.NumField(); j++ {
 				field := typ.Field(j)
 				// TODO read tags: `abi:"fieldName"`
-				if field.Name == strings.ToUpper(e.Inputs[i].Name[:1])+e.Inputs[i].Name[1:] {
-					if err := set(value.Field(j), reflectValue, e.Inputs[i]); err != nil {
+				if field.Name == strings.ToUpper(input.Name[:1])+input.Name[1:] {
+					if err := set(value.Field(j), reflectValue, input); err != nil {
 						return err
 					}
 				}
 			}
-		case reflect.Slice, reflect.Array:
-			if value.Len() < i {
```

Comment:
The slice / array size should be done once, and at the beginning. Not every time in the loop. Also the error message was wrong, because it could fail before reaching the end (reporting the wrong expected size).